src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp

```diff
@@ -594,6 +594,8 @@ class JoinOnLogicalExpressionOptimizerVisitor : public InDepthQueryTreeVisitorWi
           * Optimization that is safe for JOIN ON may not be safe for SELECT, for example:
           * Values `NULL` and `false` are not equivalent in SELECT, so we cannot change type from Nullable(UInt8) to UInt8 there, while it's valid for `JOIN ON`.
           * Also, operator <=> can be used in JOIN ON, but not in SELECT, so we need to keep original expression `a = b OR isNull(a) AND isNull(b) there.
+          *
+          * FIXME: May be removed after https://github.com/ClickHouse/ClickHouse/pull/66143
```

Comment:
This might be a stupid question and isn't strictly related to this PR, but I thought I will ask, because better to ask stupid questions than have bugs. My understanding is also not the best about how aliasing works under the hood, so it is possible that I totally misunderstood something.

Cannot be a query like `SELECT extra FROM users AS t1 JOIN users AS t2 ON t1.uid = t2.uid OR (t1.name = t2.name AS extra);` cause issues? In this case only a subclause of the `or` function is reused in the `SELECT`, but I guess we might also change subclauses. As far as I see the `clone()` before doesn't protect us against that.

Reply:
When we enter the node and see that it’s shared, we `clone` it (which do recursive clone), ensuring that any further modifications are applied to the cloned query. If we didn’t clone it earlier, it means that part of the query wasn’t shared, but only some subtree was.

Reply:
Ahh, because we don't modify any of the children, just the node that we currently entered. Understood!